src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/PreferStreamAsyncMemoryOverloads.Fixer.cs

```diff
@@ -0,0 +1,98 @@
+                if (model.GetOperation(node, ct) is IInvocationOperation invocation)
+                {
+                    string methodName = invocation.TargetMethod.Name;
+
+                    string title;
+                    if (methodName == "ReadAsync")
+                    {
+                        title = MicrosoftNetCoreAnalyzersResources.PreferStreamReadAsyncMemoryOverloadsTitle;
+                    }
+                    else if (methodName == "WriteAsync")
+                    {
+                        title = MicrosoftNetCoreAnalyzersResources.PreferStreamWriteAsyncMemoryOverloadsTitle;
+                    }
+                    else
+                    {
+                        return;
+                    }
+
+                    context.RegisterCodeFix(
+                        CodeAction.Create(
+                            title: title,
+                            createChangedDocument: c => FixInvocation(doc, root, invocation, methodName),
+                            equivalenceKey: MicrosoftNetCoreAnalyzersResources.PreferStreamAsyncMemoryOverloadsMessage),
+                        context.Diagnostics);
+                }
+            }
+        }
+
+        private static Task<Document> FixInvocation(Document doc, SyntaxNode root, IInvocationOperation invocation, string methodName)
+        {
+            SyntaxGenerator generator = SyntaxGenerator.GetGenerator(doc);
+
+            // The stream object
+            SyntaxNode instanceNode = invocation.Instance.Syntax;
+
+            // Need the byte array object so we can invoke its AsMemory() method
+            SyntaxNode bufferInstanceNode = invocation.Arguments[0].Value.Syntax; // byte[] buffer
+
+            // These arguments are not modified, just moved inside AsMemory
+            SyntaxNode offsetNode = invocation.Arguments[1].Syntax; // int offset
+            SyntaxNode countNode = invocation.Arguments[2].Syntax;  // int count
+
+            // Generate an invocation of the AsMemory() method from the byte array object
+            SyntaxNode asMemoryExpressionNode = generator.MemberAccessExpression(bufferInstanceNode, "AsMemory");
+            SyntaxNode asMemoryInvocationNode = generator.InvocationExpression(asMemoryExpressionNode, offsetNode, countNode);
+
+            // Create a new async method call for the stream object, no arguments yet
+            SyntaxNode asyncMethodNode = generator.MemberAccessExpression(instanceNode, methodName);
+
+            // Add the arguments to the async method call, with or without CancellationToken
+            SyntaxNode newInvocationExpression;
+            if (invocation.Arguments.Length > 3)
+            {
+                newInvocationExpression = generator.InvocationExpression(
+                    asyncMethodNode, asMemoryInvocationNode,
+                    invocation.Arguments[3].Syntax /* CancellationToken */);
```

Comment:
Can arguments be more then 4? The above if check indicates it can be `4 or more`. I would recommend that if the only known case here is 4 arguments, then add a bail out in RegisterCodeFixesAsync when argument count `>4` before invoking RegisterCodeFix, and change the check above to `if (invocation.Arguments.Length == 4)`

Reply:
Alternatively, can we simplify the code by completely removing the arguments length check? Something like:

```csharp
var argumentsBuilder = ImmutableArray<SyntaxNode>.CreateBuilder();
argumentsBuilder.Add(asyncMethodNode);
argumentsBuilder.Add(asMemoryInvocationNode);
foreach (var arg in invocation.Arguments.Skip(3))
{
   argumentsBuilder.Add(arg.Syntax);
}

var newInvocation = generator.InvocationExpression(argumentsBuilder.ToImmutable());
```

Reply:
The undesired methods can have either 3 or 4 arguments:
```cs
Stream.WriteAsync(Byte[], Int32, Int32)
Stream.WriteAsync(Byte[], Int32, Int32, CancellationToken)
Stream.ReadAsync(Byte[], Int32, Int32)
Stream.ReadAsync(Byte[], Int32, Int32, CancellationToken)
```

Reply:
I'll use `invocation.Arguments.Length == 4`.